### ISSUE TITLE
field creation

### ISSUE THREAD

**Issue by gregsdennis** [ietf-wg-jsonpath/draft-ietf-jsonpath-base#514]
_An interesting question was asked on [StackOverflow](https://stackoverflow.com/q/77683897/878701), and [subsequently](https://github.com/gregsdennis/json-everything/issues/602) on my lib's repo, which I thought would be a good suggestion for the V2 spec (supposing that happens)._

### Proposal

Given the JSON document

```json
{
  "options": [
    {
      "price": 217,
      "quantity": 2
    },
    {
      "price": 63,
      "quantity": 5
    }
  ]
}
```

This proposal enables performing a calculation, e.g. `price * quantity`, and getting its result by assigning a new property in the local JSON so that it can be queried by subsequence segments.

A path that could do this would be:

```jsonpath
$.options[?(@.result = @.price * @.quantity)].result
```

In this path, `@.result` is assigned the result of the multiplication.  This assignment would operate as an "upsert" where it would overwrite any value that may already be there.  The new value is subsequently returned by the next segment `.result`.

This likely has applications beyond this simple example and is worth considering.  Additionally, many other query languages (e.g. SQL) permit the creation and selection of temporary values like this.

It's understood that filter expressions as currently defined are required to return a boolean value, and the above expression does not, so that aspect of filter expressions will need to be revisited/addressed.

Depending on how the conversation goes here, I could probably work it up in my implementation as an optional behavior to make it available on [my playground](https://json-everything.net/json-path).

### Caveat

This feature is predicated on math operations being supported, which the base spec does not.  Currently we can't even select based on calculated values meeting some threshold, e.g.

```jsonpath
$.options[?(@.price * @.quantity > 400)]
```

Note that these math operators (not assignment, though) _are_ already optionally supported on my playground.

**Comment by cabo** [comment]
Hi Greg,

in relational algebra, the ability to select a subset of the columns of the rows is called projection.
In SQL, projection confusingly happens in the the SELECT part of a query.
SQL SELECT adds the ability to *add* columns (e.g., by computing them) to projection, quite similar to what you are proposing (but then you are missing the original projection that is about selecting columns).

We decided to limit JSONPath-base to selection (confusingly the part that happens in the WHERE clause in SQL).  We may want to address projection in a future extension.  If yes, I think we should do this in a simple but powerful way (e.g., including the ability to cut down on columns, not just adding them).  I think your proposal is a way to experiment with projection.  Coming up with an expression language that can do more powerful operations on values than JSONPath can do today will be another component of such a proposal.  A way to present results that include values that are not in the original query argument will be another.

**Comment by gregsdennis** [comment]
I think projections in general would be a very appreciated feature that could potentially lead to something like transformations (which I have also seen requested).

**Comment by danielaparker** [comment]
> I think projections in general would be a very appreciated feature that could potentially lead to something like transformations (which I have also seen requested).

@gregsdennis I don't think it makes sense to talk about projections in JSONPath as if they had anything to do with projections in SQL. SQL is based on relational algebra, which provides its theoretical foundation. JSONPath has no such foundation, it's an ad hoc construction. Relational algebra is closed under the operators, that is, the input to any operator is a relation, and the output is a relation, which allows expressions to be nested in an arbitrary manner. Expressions in IETF JSONPath do not appear to be closed under anything.

Regarding transformations, it's not obvious why you would want to introduce that into JSONPath, as there are other well established JSON query languages that have that, notably [JMESPath](https://jmespath.org/), [JSONAta](https://jsonata.org/), and [XPATH 3.1](https://www.w3.org/TR/xpath-31/). JSONPath's unique feature is that it returns nodes that correspond to location/value pairs in the original document, which has its own utility, particularly in implementations that support update in place in the original document.

**Comment by danielaparker** [comment]
> In this path, `@.result` is assigned the result of the multiplication. This assignment would operate as an "upsert" where it would overwrite any value that may already be there. The new value is subsequently returned by the next segment `.result`.
> 
It sounds to me like this would be a use case for lexical scoping. JMESPath recently accepted a [JMESPath Enhancement Proposal](https://jmespath.github.io/jmespath.jep/0018-lexical-scope.html) for lexical scoping using a new let expression.

**Comment by gregsdennis** [comment]
> JSONPath's unique feature is that it returns nodes that correspond to location/value pairs in the original document...

It's actually a good point that JSON Path returns a set of _nodes_, not values.  There is no node that corresponds to the output of

```jsonpath
$.options[?(@.result = @.price * @.quantity)].result
```

Since `result` is a value that's generated at evaluation time, there's no path that leads to it.

**Comment by danielaparker** [comment]
> Since `result` is a value that's generated at evaluation time, there's no path that leads to it.

Indeed. It should be noted though that the two biggest JSONPath implementations - Jayway JSONPath and JSONPath Plus - do produce new values in some circumstances. Jayway JSONPath supports path functions , e.g. `$.phoneNumbers.length()`, that returns the number of elements in an array as a new value, and JSONPath Plus supports the length property, e.g. `$.phoneNumbers.length`, that does the same.  Jayway JSONPath raises an error if attempting to return normalized paths when using path functions.
